design-proposals/configurable-features.md

```diff
@@ -0,0 +1,77 @@
+# Overview
+With the introduction of [KubeVirt Feature Lifecycle](https://github.com/kubevirt/community/blob/main/design-proposals/feature-lifecycle.md) policy, features reaching General Availability (GA) need to drop their use of feature gates. This applies also to configurable features that we may want to disable.
+
+## Motivation
+Users or developers may want certain features to be in a given state, for example to make the best use out of given resources or for compliance reasons features may expose sensitive information from the host to the virtual machines (VM) or add additional containers to the launcher pod, which are not required by the user. The behavior of other features might be changed by editing configurables, e.g. the maximum of CPU sockets allowed for each VM can be configured.
+
+Before the introduction of [KubeVirt Feature Lifecycle](https://github.com/kubevirt/community/blob/main/design-proposals/feature-lifecycle.md) policy, many feature gates remained after feature's graduation to GA with the sole purpose of acting as a switch for the feature. Generally speaking, this is a bad practice, because feature gates should be reserved for controlling a feature until it reaches maturity. i.e., GA. Therefore, in the case that a developer wants to provide the ability to tune/change the state of the feature, configurables exposed in the KubeVirt CR should be provided. Ideally, this should be accomplished while achieving [eventually consistency](https://en.wikipedia.org/wiki/Eventual_consistency). This forces us to avoid the feature state control checking on webhooks and moving the feature state control closer to the responsible code. Moreover, it has to be decided how the system should behave if a virtual machine instance (VMi) is requiring a feature in a state different from what was configured in the KubeVirt CR, or what should happen if the configuration of a feature in use is changed. (see matrix below).
+
+## Goals
+- Establish how the features enablement switch should work.
+- Describe how the system should react in these scenarios:
+    - A feature in KubeVirt is set to state A and a VMi requests the feature to be in state B.
+    - A feature in KubeVirt is set to state A, there are running VMis using the feature in state A, and the feature is changed in KubeVirt to state B.
+    - A feature in KubeVirt is set to state A, and pending VMis want to use it.
+    - A feature in KubeVirt is set to state A, and running VMis using the feature in state B wants to live migrate.
+- Graduate as many features as possible from features gates to configurables.
+
+## Non Goals
+- Describe how features protected with features gates should work.
+
+## Definition of Users
+Development contributors.
+
+Cluster administrators.
+
+## User Stories
+As a developer, I want to make a given feature configurable.
```

Comment:
```suggestion
As a VM owner, I want to use a given feature.
```

I don't know if VM owners really want to configure a feature, they just want to use it.

Reply:
I've added the "VM owner" use case. However, IMHO the use case in which the "As a developer, I want..." is valid. This design proposal should provide the guidelines for developers to create feature configurables. WDYT?

Reply:
I'm not sure who is the "developer" persona. Do you refer to a VM owner that want to develop and application inside it?

Can you give an example of a feature that a developer would like to "configure" instead of just "use in their VM".

Reply:
I meant the KubeVirt developer itself. For example, I'm the owner of the downward metrics feature, and me, as the developer, I want to make this feature configurable, i.e, provide a mechanism to users to enable and disable the feature cluster-wide.

Reply:
The purpose of the "User Stories" section of the proposal is to disassociate us developers a bit. In a user story, you need to put yourself in the shoes of the end user, either the cluster-admin or a VM owner, and think - why would that person want to configure this feature? Isn't it better for them to have it always on?

There may be features that are directed at KubeVirt developers, intended to help testing or debug, but typically they are few. Most of the code we write is for others and we should have them in mind.

Reply:
Got it! Thanks for the wide explanation. I've dropped the "developer" user history.